### ISSUE TITLE
Support SharedCredentials in AWS Provider

### ISSUE THREAD

**Issue by tischler** [hashicorp/terraform#2235]
AWS provider should have support for shared credentials.  The configuration would look something like: 
```
provider "aws" {
  profile = "aws-staging"
  region = "us-west-1"
}
```
or
```
provider "aws" {
  profile = "aws-staging"
  filename = "../aws-credentials"
  region = "us-west-1"
}
```

It appears that the credential chain in builtin/providers/aws/config.go already exists, the task should be: 
1. Export 'profile' and 'credentials_filename' parameters in aws/provider.go
2. Make the current 'access_key' and 'secret_key' optional. (is there a way to declare ORs in required parameters?)
3. Add the new parameters to the NewChainCredentials SharedCredentialsProvider
4. Update documentation.

**Comment by radeksimko** [comment]
Hi @tischler 
see https://github.com/hashicorp/terraform/issues/1822

Is it describing all you're asking for? If not, maybe we can modify the original issue instead, unless you feel there's a reason to keep it separate.

**Comment by tischler** [comment]
these could very well be done with the same work effort, or they could be separate.  I have a number of different profiles in my shared creds and I want to have the default aws_account/aws_secret to be optional, and let me specify which exact profile I want.  I would be bypassing caring about the order... (before using terraform, I didn't export an environment variable at all).

**Comment by jjshoe** [comment]
#266 looks related to me.

**Comment by jeekajoo** [comment]
`packer` has this feature and it is able to read creds from awscli config (~/.aws/config or ~/.aws/credentials)
```
AWS_PROFILE=myprofile packer build packer.json
```

**Comment by scarolan** [comment]
+1 on this.  If terraform could somehow support ~/.aws/credentials or ~/.aws/config that would make life easier.

**Comment by scarolan** [comment]
Just in case anyone else runs across this and is confused by how to use environment variables to define your AWS credentials with terraform:

1.  Make sure you have exported AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY via a dotfile or script.
2.  Make your example.tf and terraform.tfvars files look like these:

https://gist.github.com/scarolan/60ae8a2d5f2a8fdb5c55

**Comment by jen20** [comment]
This has been implemented in #4632 and should hit in the next release of Terraform.
